File: sdk/java/src/test/java/dev/rue/hook/ServeTest.java
package dev.rue.hook;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** The loop: one reply per request line, nothing for anything else, and a budget. */
class ServeTest {

    private static List<Map<String, Object>> serve(String input, Hooks hooks, Duration budget) {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        PrintStream out = new PrintStream(bytes, true, StandardCharsets.UTF_8);
        Serve.serve(new BufferedReader(new StringReader(input)), out, hooks, budget);
        List<Map<String, Object>> replies = new ArrayList<>();
        for (String line : bytes.toString(StandardCharsets.UTF_8).split("\n")) {
            if (!line.isBlank()) replies.add(Json.parseObject(line));
        }
        return replies;
    }

    private static Hooks probe(long sleepMs) {
        Hooks h = new Hooks();
        h.probe = (host, p) -> {
            try {
                Thread.sleep(sleepMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return Hooks.observation("yes", "ok");
        };
        return h;
    }

    @Test
    void onlyRequestsAreAnsweredAndEachOnce() {
        String input = String.join("\n",
                "",
                "not json at all",
                "[\"an\",\"array\"]",
                "{\"event\":{\"plan\":\"p\"}}",
                "{\"id\":1}",
                "{\"id\":2,\"kind\":\"probe\",\"op\":\"observe\",\"host\":\"h\",\"probe\":\"up\"}",
                "[" .repeat(50_000),
                "{\"id\":3,\"kind\":\"probe\",\"op\":\"observe\",\"host\":\"h\",\"probe\":\"up\"}");
        List<Map<String, Object>> replies = serve(input, probe(0), null);
        assertEquals(2, replies.size(), String.valueOf(replies));
        assertEquals(2L, replies.get(0).get("id"));
        assertEquals(3L, replies.get(1).get("id"), "a pathological line did not end the loop");
    }

    @Test
    void aHandlerOverItsBudgetAnswersNoAndSaysWhy() {
        String line = "{\"id\":9,\"kind\":\"probe\",\"op\":\"observe\",\"host\":\"h\",\"probe\":\"up\"}\n";
        Map<String, Object> slow = serve(line, probe(150), Duration.ofMillis(20)).get(0);
        assertEquals(9L, slow.get("id"));
        assertEquals(false, slow.get("ok"));
        assertTrue(String.valueOf(slow.get("error")).contains("budget"), String.valueOf(slow));
        Map<String, Object> quick = serve(line, probe(0), Duration.ofSeconds(5)).get(0);
        assertEquals(true, quick.get("ok"));
    }

    @Test
    void aReplyThatCannotBeWrittenBecomesARefusalNotADeadHook() {
        Hooks h = new Hooks();
        h.probe = (host, p) -> Map.of("text", "", "tri", Double.NaN);
        String input = "{\"id\":5,\"kind\":\"probe\",\"op\":\"observe\",\"host\":\"h\",\"probe\":\"x\"}\n"
                + "{\"id\":6,\"kind\":\"probe\",\"op\":\"observe\",\"host\":\"h\",\"probe\":\"x\"}\n";
        List<Map<String, Object>> replies = serve(input, h, null);
        assertEquals(2, replies.size(), String.valueOf(replies));
        assertEquals(false, replies.get(0).get("ok"));
        assertEquals(6L, replies.get(1).get("id"));
    }
}

File: sdk/elixir/test/serve_test.exs
defmodule RueHook.ServeTest do
  # The loop: one reply per request line, nothing for anything else, a
  # budget, and no line or reply that ends the hook.
  use ExUnit.Case, async: true

  alias RueHook.{Hooks, Serve}

  defmodule Slow do
    def observe(_host, "slow"), do: (Process.sleep(150); {:ok, RueHook.observation(:yes, "ok")})
    def observe(_host, "tuple"), do: {:ok, %{"text" => {:not, :json}, "tri" => "yes"}}
    def observe(_host, _), do: {:ok, RueHook.observation(:yes, "ok")}
  end

  defp req(id, probe \\ "up"),
    do: JSON.encode!(%{"id" => id, "kind" => "probe", "op" => "observe", "host" => "h", "probe" => probe})

  defp serve(lines, budget_ms \\ nil) do
    {:ok, input} = StringIO.open(Enum.join(lines, "\n") <> "\n")
    {:ok, output} = StringIO.open("")
    :ok = Serve.pump(input, output, %Hooks{probe: Slow}, budget_ms)
    {_, written} = StringIO.contents(output)
    written |> String.split("\n", trim: true) |> Enum.map(&JSON.decode!/1)
  end

  test "only requests are answered, and each once" do
    replies = serve(["", "not json", ~s(["an","array"]), ~s({"event":{"plan":"p"}}), ~s({"id":1}), req(2), req(3)])
    assert Enum.map(replies, & &1["id"]) == [2, 3]
  end

  test "a handler over its budget answers no and says why" do
    [slow] = serve([req(9, "slow")], 20)
    assert {slow["id"], slow["ok"]} == {9, false}
    assert slow["error"] =~ "budget"
    [quick] = serve([req(9)], 5_000)
    assert quick["ok"] == true
  end

  test "a reply that cannot be written becomes a refusal, not a dead hook" do
    replies = serve([req(5, "tuple"), req(6)])
    assert Enum.map(replies, & &1["id"]) == [5, 6]
    assert hd(replies)["ok"] == false
  end
end
